go/libraries/doltcore/remotestorage/writebuffer.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2025 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remotestorage
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/dolthub/dolt/go/store/hash"
+	"github.com/dolthub/dolt/go/store/nbs"
+)
+
+type WriteBuffer interface {
+	Put(nbs.CompressedChunk) error
+
+	// Returns the current set of written chunks.  After this
+	// returns, concurrent calls to other methods may block until
+	// |WriteCompleted is called.  Calls to |GetAllForWrite| must
+	// be bracketed by a call to |WriteCompleted|
+	GetAllForWrite() map[hash.Hash]nbs.CompressedChunk
+
+	// Called after a call to |GetAllForWrite|, this records
+	// success or failure of the write operation.  If the write
+	// operation was successful, then the written chunks are now
+	// in the upstream, and so they can be cleared. Otherwise, the
+	// written chunks are retained in the write buffer so that the
+	// write can be retried.
+	WriteCompleted(success bool)
+
+	// ChunkStore clients expect to read their own writes before a commit.
+	// On the get path, remotestorage should add pending chunks to its result
```

Comment:
Not clear initially that `res` is the result set. Would be clearer with:

```suggestion
	// On the get path, remotestorage should add pending chunks to |res|
```